lib/offsite_payments/integrations/migs.rb

```diff
@@ -0,0 +1,231 @@
+          options = @options
+          builder = TransactionBuilder.new(@credentials)
+          builder.add_invoice(options)
+          builder.add_creditcard_type(options[:card_type]) if options[:card_type]
+          builder.add_amount(cents)
+          builder.add_standard_parameters('pay', options[:unique_id])
+          post = builder.post.merge(
+            :Locale => options[:locale] || 'en',
+            :ReturnURL => options.fetch(:return_url)
+          )
+          post[:SecureHash] = SecureHash.calculate(@secure_hash, post)
+
+          self.class.parent.base_url + '?' + post_data(post)
+        end
+
+        private
+
+        def post_data(post)
+          post.collect { |key, value| "vpc_#{key}=#{CGI.escape(value.to_s)}" }.join("&")
+        end
+      end
+
+      class Notification < OffsitePayments::Notification
+        def initialize(params, options = {})
+          @params = params
+          @response = parse
+          @options = options
+        end
+
+        def parse
+          hash = {}
+          @params.each do |key, value|
+            hash[key.gsub('vpc_', '').to_sym] = value
+          end
+          hash
+        end
+
+        def avs_response_code
+          avs_response_code = @response[:AVSResultCode]
+          avs_response_code = 'S' if avs_response_code == "Unsupported"
+          avs_response_code
+        end
+
+        def cvv_result_code
+          cvv_result_code = @response[:CSCResultCode]
+          cvv_result_code = 'P' if cvv_result_code == "Unsupported"
+          cvv_result_code
+        end
+
+        def success?
+          @response[:TxnResponseCode] == '0'
+        end
+
+        def fraud_review?
+          ISSUER_RESPONSE_CODES[@response[:AcqResponseCode]] == 'Suspected Fraud'
+        end
+
+        def acknowledge
+          # Failures don't include a secure hash, so return directly
+          return true unless success?
+
+          # Check SecureHash only on success (not returned otherwise)
+          expected_secure_hash = SecureHash.calculate(@options[:secure_hash], @response)
+          unless @response[:SecureHash] == expected_secure_hash
+            raise SecurityError, "Secure Hash mismatch, response may be tampered with"
+          end
+
+          true
+        end
+
+        def gross
+          @response[:Amount].to_d / 100.0
+        end
+
+        def card_code
+          return unless @response.key?(:Card) # Card doesn't appear on failure
+          migs_code = @response[:Card]
+          CARD_TYPES.detect { |ct|
+            ct.migs_code == migs_code
+          }.am_code
+        end
+
+        def order_id
+          @response[:OrderInfo]
+        end
+
+        def uid
+          @response[:MerchTxnRef]
+        end
+
+        def transaction_id
+          @response[:TransactionNo]
+        end
+
+        def message # only when error
+          @response['Message']
+        end
+      end
+
+      def test?
+        @options[:login].start_with?('TEST')
```

Comment:
I have commented to explain TEST is the prefix used for test systems. 
This is beyond our control, unless you think another aspect of this is 
bad, in which case please clarify and ideally suggest an improvement so 
we can discuss.

On 12/01/16 13:17, Alessandro Berardi wrote:
>
> In lib/offsite_payments/integrations/migs.rb 
> <https://github.com/sealink/offsite_payments/pull/1#discussion_r49410226>:
>
> > +
> > +        def uid
> > +          @response[:MerchTxnRef]
> > +        end
> > +
> > +        def transaction_id
> > +          @response[:TransactionNo]
> > +        end
> > +
> > +        def message # only when error
> > +          @response['Message']
> > +        end
> > +      end
> > +
> > +      def test?
> > +        @options[:login].start_with?('TEST')
>
> Not sure this is good/safe/nice.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/sealink/offsite_payments/pull/1/files#r49410226>.
>
>
> ______________________________________________________________________
> This email has been scanned by the Symantec Email Security.cloud service.
> For more information please visit http://www.symanteccloud.com
> ______________________________________________________________________